File: internal/localcompile/keygen.go
package localcompile

import (
	"crypto/ecdh"
	"crypto/rand"
	"encoding/base64"
	"fmt"

	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
)

// wgKeyLen is the X25519 key length in bytes (private and public). It mirrors
// wgtypes.KeyLen; declared here so the stdlib ecdhKeygen path does not import a
// constant for it.
const wgKeyLen = 32

// wgtypesKeygen is the default Keygen: it wraps today's exact wgtypes calls so the
// production pipeline is byte-for-byte unchanged. It is the implementation
// CompileRequest.Keygen defaults to (nil ⇒ wgtypesKeygen) and the byte-identity
// oracle the equivalence test pins ecdhKeygen against.
//
// It satisfies both the localcompile.Keygen contract member and render.Keygen
// (the same method set declared on the render consumer side — render must not
// import this package, so the seam interface is duplicated there by design).
type wgtypesKeygen struct{}

var _ Keygen = wgtypesKeygen{}

// DerivePublic reproduces wgtypes.ParseKey(privB64).PublicKey().String().
func (wgtypesKeygen) DerivePublic(privB64 string) (string, error) {
	priv, err := wgtypes.ParseKey(privB64)
	if err != nil {
		return "", err
	}
	return priv.PublicKey().String(), nil
}

// Generate reproduces wgtypes.GeneratePrivateKey() + .String() / .PublicKey().String().
func (wgtypesKeygen) Generate() (string, string, error) {
	priv, err := wgtypes.GeneratePrivateKey()
	if err != nil {
		return "", "", err
	}
	return priv.String(), priv.PublicKey().String(), nil
}

// ParseAndNormalize reproduces wgtypes.ParseKey(privB64).String() — the air-gap
// case-a re-canonicalization that is persisted back onto the node. wgtypes does
// NOT clamp on parse/String, so this is a length-validated base64 round-trip.
func (wgtypesKeygen) ParseAndNormalize(privB64 string) (string, error) {
	priv, err := wgtypes.ParseKey(privB64)
	if err != nil {
		return "", err
	}
	return priv.String(), nil
}

// ecdhKeygen is the stdlib reference Keygen: it derives X25519 public keys through
// crypto/ecdh instead of wgtypes/wgctrl, giving plan-5 (conformance) and plan-4
// (the TypeScript port) a wgctrl-free, stdlib-anchored definition of the key
// derivation and unblocking a future js/wasm Go oracle. It is proven byte-equal to
// wgtypesKeygen over 10k inputs (keygen_equivalence_test.go) on DerivePublic and
// ParseAndNormalize.
//
// It is a REFERENCE/TEST implementation: it is deliberately NOT wired into any
// production caller (production defaults to wgtypesKeygen).
type ecdhKeygen struct{}

var _ Keygen = ecdhKeygen{}

// DerivePublic decodes the base64 private key, derives the public key via
// crypto/ecdh X25519 (which clamps the scalar internally, exactly as
// curve25519.ScalarBaseMult does inside wgtypes' Key.PublicKey), and re-encodes.
func (ecdhKeygen) DerivePublic(privB64 string) (string, error) {
	raw, err := decodeKey(privB64)
	if err != nil {
		return "", err
	}
	priv, err := ecdh.X25519().NewPrivateKey(raw)
	if err != nil {
		return "", fmt.Errorf("ecdh: derive public from X25519 private key: %w", err)
	}
	return base64.StdEncoding.EncodeToString(priv.PublicKey().Bytes()), nil
}

// Generate produces a fresh X25519 key pair using the same clamp wgtypes applies
// in GeneratePrivateKey (key[0]&=248; key[31]&=127; key[31]|=64), then derives the
// public half via crypto/ecdh. The private key is random, so its bytes are never
// byte-asserted; only that the derived public matches what DerivePublic would
// yield for the same private key.
func (ecdhKeygen) Generate() (string, string, error) {
	raw := make([]byte, wgKeyLen)
	if _, err := rand.Read(raw); err != nil {
		return "", "", fmt.Errorf("ecdh: read random bytes: %w", err)
	}
	// Same X25519 clamp as wgtypes.GeneratePrivateKey (https://cr.yp.to/ecdh.html).
	raw[0] &= 248
	raw[31] &= 127
	raw[31] |= 64

	priv, err := ecdh.X25519().NewPrivateKey(raw)
	if err != nil {
		return "", "", fmt.Errorf("ecdh: build X25519 private key: %w", err)
	}
	privB64 := base64.StdEncoding.EncodeToString(raw)
	pubB64 := base64.StdEncoding.EncodeToString(priv.PublicKey().Bytes())
	return privB64, pubB64, nil
}

// ParseAndNormalize round-trips a private key to its canonical base64 form. It
// mirrors wgtypes' ParseKey(...).String(), which is a length-validated base64
// round-trip with NO clamping, so this decodes (validating the 32-byte length) and
// re-encodes without going through ecdh — clamping here would diverge from
// wgtypes' String() and corrupt the air-gap case-a re-write byte-for-byte.
func (ecdhKeygen) ParseAndNormalize(privB64 string) (string, error) {
	raw, err := decodeKey(privB64)
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(raw), nil
}

// decodeKey base64-decodes a key and validates the X25519 length, matching the two
// failure modes of wgtypes.ParseKey (bad base64, wrong length).
func decodeKey(b64 string) ([]byte, error) {
	raw, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		return nil, fmt.Errorf("decode base64 X25519 key: %w", err)
	}
	if len(raw) != wgKeyLen {
		return nil, fmt.Errorf("X25519 key has length %d, want %d", len(raw), wgKeyLen)
	}
	return raw, nil
}

File: frontend/src/compiler/keygen.ts
// WireGuard key-derivation seam — the TypeScript mirror of the plan-3-frozen three-operation Keygen
// interface (render.go:66-78). This is the browser-side replacement for the Go wgtypesKeygen
// (render.go:86-114), which wraps wgtypes.ParseKey/.PublicKey()/.String(). The three operations are:
//
//   - derivePublic(privB64)      — wgtypes.ParseKey(priv).PublicKey().String() (render.go:88-94)
//   - generate()                 — wgtypes.GeneratePrivateKey() + .String()/.PublicKey().String()
//                                   (render.go:97-104)
//   - parseAndNormalize(privB64) — wgtypes.ParseKey(priv).String() (render.go:106-113), the air-gap
//                                   case-a re-canonicalization persisted back onto the node.
//
// X25519 derivation goes through @noble/curves x25519.getPublicKey — the DOCUMENTED API, which clamps
// the scalar INTERNALLY (key[0] &= 248; key[31] &= 127; key[31] |= 64) exactly as crypto/ecdh and
// wgtypes do. We do NOT pre-clamp and do NOT call scalarMultBase: a hand-clamp before a getPublicKey
// that also clamps risks a double-clamp divergence. Clamp/derivation correctness is pinned by
// keygen.kat.test.ts against the Go-authored RFC 7748 + clamp-boundary vectors, not by this prose.
//
// All base64 here is STANDARD (with padding), NOT url — mirroring Go's base64.StdEncoding, which
// wgtypes.ParseKey decodes and Key.String() encodes (wgtypes/types.go:126/154). webauthn.ts's base64
// helpers are RawURLEncoding (url alphabet, no padding) and MUST NOT be reused here.
//
// The seam intentionally throws PLAIN Errors on bad input, mirroring wgtypes.ParseKey returning a bare
// error. The consumer (render.go GenerateKeys, :170-218) is what wraps that into apierr codes
// (keygen_privkey_parse_failed, keygen_pinned_pubkey_no_privkey, keygen_generate_failed) — that
// wrapping is a later-phase compile concern, NOT this leaf's responsibility. The public-only-no-private
// hard error (render.go case-b, :209) likewise lives in the compile pass, not here.

import { x25519 } from '@noble/curves/ed25519.js';

// KEY_LEN is the WireGuard key length in bytes (wgtypes.KeyLen = 32, wgtypes/types.go:72). Both private
// and public X25519 keys are exactly this many raw bytes.
const KEY_LEN = 32;

// base64StdDecode decodes a standard-base64 string to raw bytes, FAITHFULLY to Go's
// base64.StdEncoding.DecodeString — NOT just "atob, which raises on malformed input". atob is too
// lenient on two axes where it would silently accept a key Go REJECTS: it tolerates unpadded base64
// and (in some engines) embedded whitespace such as space/tab. Go's StdEncoding decoder, by contrast,
// IGNORES embedded '\n' and '\r' (the only whitespace it skips) but REJECTS space, tab, and every other
// non-alphabet byte, and REQUIRES correct '=' padding so the encoded length is a multiple of 4. Verified
// against base64.StdEncoding directly: a space, a tab, and an unpadded 32-byte key are all rejected;
// embedded '\n'/'\r' are accepted.
//
// So we reproduce that grammar exactly: strip ONLY '\n' and '\r' (the bytes Go's decoder skips), then
// require the remainder to be standard-alphabet chars with 0–2 trailing '=' and a length that is a
// multiple of 4 — rejecting anything else (space, tab, unpadded, mid-string padding) BEFORE handing the
// cleaned string to atob. That makes this byte-exact to Go in BOTH directions (no TS-accepts/Go-rejects
// and no TS-rejects/Go-accepts divergence). It throws on a malformed string, matching wgtypes.ParseKey's
// "failed to parse base64-encoded key" error path (wgtypes/types.go:128).
const STD_BASE64_RE = /^[A-Za-z0-9+/]*={0,2}$/;

function base64StdDecode(s: string): Uint8Array {
  // Go's StdEncoding decoder skips embedded '\n' and '\r' (and ONLY those); strip them first.
  const cleaned = s.replace(/[\n\r]/g, '');
  if (cleaned.length % 4 !== 0 || !STD_BASE64_RE.test(cleaned)) {
    throw new Error('illegal base64 data');
  }
  const bin = atob(cleaned);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) {
    out[i] = bin.charCodeAt(i);
  }
  return out;
}

// base64StdEncode encodes raw bytes to a standard-base64 string (with padding), mirroring Go's
// base64.StdEncoding.EncodeToString. String.fromCharCode over the byte values builds the latin1 string
// btoa expects; btoa emits the standard alphabet (+/ and = padding) — byte-identical to Go's
// Key.String() (wgtypes/types.go:154).
function base64StdEncode(bytes: Uint8Array): string {
  let bin = '';
  for (let i = 0; i < bytes.length; i++) {
    bin += String.fromCharCode(bytes[i]);
  }
  return btoa(bin);
}

// parsePrivateKey decodes a base64-std private key and validates it is exactly 32 bytes, mirroring
// wgtypes.ParseKey (wgtypes/types.go:125-133): decode, then reject any length != KeyLen. The two error
// shapes (bad base64 vs wrong size) match the wgtypes branches; the consumer surfaces them as
// keygen_privkey_parse_failed with the detail attached.
function parsePrivateKey(privB64: string): Uint8Array {
  const raw = base64StdDecode(privB64);
  if (raw.length !== KEY_LEN) {
    throw new Error(`wgtypes: incorrect key size: ${raw.length}`);
  }
  return raw;
}

// derivePublic returns the base64-std public key for a base64-std X25519 private key. Mirrors
// wgtypesKeygen.DerivePublic = wgtypes.ParseKey(priv).PublicKey().String() (render.go:88-94): parse +
// validate 32 bytes, X25519 scalar-base-mult (getPublicKey clamps internally), re-encode std base64.
export function derivePublic(privB64: string): string {
  const priv = parsePrivateKey(privB64);
  const pub = x25519.getPublicKey(priv);
  return base64StdEncode(pub);
}

// generate returns a fresh (priv, pub) X25519 key pair, both base64-std. Mirrors
// wgtypesKeygen.Generate = wgtypes.GeneratePrivateKey() + .String()/.PublicKey().String()
// (render.go:97-104, the air-gap case-c brand-new pair). crypto.getRandomValues fills 32 cryptographic
// random bytes for the private scalar; getPublicKey derives the public half (clamping internally). The
// local path never emits the AgentHeld PrivateKeyPlaceholder — that custody mode is server-only.
export function generate(): { priv: string; pub: string } {
  const priv = new Uint8Array(KEY_LEN);
  crypto.getRandomValues(priv);
  const pub = x25519.getPublicKey(priv);
  return { priv: base64StdEncode(priv), pub: base64StdEncode(pub) };
}

// parseAndNormalize round-trips a base64-std private key to its canonical base64-std form. Mirrors
// wgtypesKeygen.ParseAndNormalize = wgtypes.ParseKey(priv).String() (render.go:106-113): decode +
// validate 32 bytes, then re-encode the SAME raw bytes with standard base64. This re-canonicalizes a
// non-canonical-but-valid encoding to wgtypes' exact String() form, which the air-gap case-a write-back
// persists onto the node. The ONLY non-canonical input Go's StdEncoding (and hence this decoder) accepts
// is one carrying embedded '\n'/'\r', which the decoder skips before re-encoding to the newline-free
// canonical form; an unpadded or space/tab-laden variant is REJECTED (it is not valid StdEncoding), so it
// never reaches this round-trip. It is a pure encoding round-trip — it does NOT clamp or otherwise mutate
// the key bytes.
export function parseAndNormalize(privB64: string): string {
  const raw = parsePrivateKey(privB64);
  return base64StdEncode(raw);
}
